File: DSA/SparshJain/Day_9_Q1.cpp
#include <stack>
#include <iostream>
#include <algorithm>

class MaxStack {
private:
    std::stack<int> stack;
    std::stack<int> max_stack;

public:
    void push(int x) {
        stack.push(x);
        if (max_stack.empty() || x >= max_stack.top()) {
            max_stack.push(x);
        }
    }

    void pop() {
        if (!stack.empty()) {
            if (stack.top() == max_stack.top()) {
                max_stack.pop();
            }
            stack.pop();
        }
    }

    int get_max() {
        if (!max_stack.empty()) {
            return max_stack.top();
        }
        return -1; 
    }
};

int main() {
    MaxStack stack;
    stack.push(1);
    stack.push(2);
    std::cout << stack.get_max() << std::endl; 
    stack.pop();
    std::cout << stack.get_max() << std::endl; 
  return 0;
}

File: DSA/Aanidhay-Aggarwal/Day-9/Day9_q1.java
import java.util.Stack;

public class MaxStack {
    private Stack<Integer> stack;
    private Stack<Integer> maxStack;

    public MaxStack() {
        stack = new Stack<>();
        maxStack = new Stack<>();
    }

    public void push(int x) {
        stack.push(x);
        if (maxStack.isEmpty() || x >= maxStack.peek()) {
            maxStack.push(x);
        }
    }

    public int pop() {
        if (stack.isEmpty()) {
            return -1;
        }
        int top = stack.pop();
        if (top == maxStack.peek()) {
            maxStack.pop();
        }
        return top;
    }

    public int getMax() {
        if (maxStack.isEmpty()) {
            return -1;  // or throw an exception
        }
        return maxStack.peek();
    }

    public static void main(String[] args) {
        MaxStack stack = new MaxStack();
        stack.push(1);
        stack.push(2);
        System.out.println(stack.getMax());
        stack.pop();
        System.out.println(stack.getMax());
    }
}
